keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md

```diff
@@ -0,0 +1,478 @@
+sufficient free hugepages exist.
+
+## Motivation
+
+The Memory Manager tracks hugepage allocations for Guaranteed QoS pods to provide
+NUMA-aware memory and hugepage pinning. However, it operates on its internal
+accounting without verifying the actual state of hugepages on the system.
+
+This creates a problem when:
+1. Burstable or BestEffort pods consume hugepages (via hugetlbfs mounts or
+   `mmap` with `MAP_HUGETLB`) without being tracked by the Memory Manager
+2. External processes or other system components consume hugepages
+3. The Memory Manager's internal state becomes stale or inconsistent with reality
+
+In these scenarios, a Guaranteed pod requesting hugepages may be admitted based
+on the Memory Manager's internal tracking, only to fail at runtime when the
+container attempts to use the already-exhausted hugepages.
+
+### Goals
+
+- Verify OS-reported free hugepages during pod admission for the Static policy
+- Reject pods requesting hugepages when insufficient free hugepages are available
+- Provide clear error messages when admission fails due to insufficient hugepages
+- Maintain backwards compatibility with existing Memory Manager behavior
+
+### Non-Goals
+
+- Track hugepage usage by Burstable or BestEffort pods in the Memory Manager
+- Modify scheduler behavior or add hugepage awareness to the scheduler
+- Provide hugepage reservation or preemption mechanisms
+- Support platforms other than Linux
+
+## Proposal
+
+Enhance the Memory Manager's Static policy to verify actual hugepage availability
+by querying sysfs during pod admission. This involves:
+
+1. **cadvisor enhancement**: Add a `FreePages` field to `HugePagesInfo` struct
+   that reports free hugepages per NUMA node, read from sysfs
+
+2. **Memory Manager enhancement**: During `Allocate()` in the Static policy,
+   verify that OS-reported free hugepages meet or exceed the requested amount
+   before admitting the pod
+
+### User Stories
+
+#### Story 1: DPDK Application Admission Failure
+
+As a cluster administrator running DPDK-based network functions, I deploy a
+Burstable pod that mounts hugetlbfs and consumes 2GB of 1GB hugepages for packet
+buffer pools. Later, I deploy a Guaranteed pod also requesting 2GB of 1GB hugepages.
+
+**Current behavior**: The Guaranteed pod is admitted (Memory Manager shows
+hugepages as available) but fails at container startup when DPDK tries to allocate
+hugepages that are already consumed.
+
+**Desired behavior**: The Guaranteed pod admission fails immediately with a clear
+error indicating insufficient free hugepages, allowing the scheduler to try
+another node or the administrator to take corrective action.
+
+#### Story 2: Database Workload with Hugepages
+
+As a database administrator, I run PostgreSQL with hugepages enabled for shared
+buffers. If an external monitoring agent or debugging tool temporarily consumes
+hugepages, subsequent Guaranteed pods requesting hugepages should not be admitted
+until hugepages are freed.
```

Comment:
Arguably, a fix can be running these agents or tools as pod, so kubelet and kubernetes and the platform in general are aware of their presence and the accounting is correct. Even if we add the extra checks and more precise accounting, which I would generally in favor of, unexpected rejections are usually confusing the users.